internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,479 @@
+	for _, ping := range pingMap {
+		if ping.request == nil { // this should not happen
+			return errors.New("internal error: ping.request is nil")
+		}
+		if len(ping.responses) == 0 {
+			tk.Pings = append(tk.Pings, &SinglePing{
+				ConnIdDst:   ping.request.dstID,
+				ConnIdSrc:   ping.request.srcID,
+				Failure:     archival.NewFailure(timeoutErr),
+				Request:     &model.ArchivalMaybeBinaryData{Value: string(ping.request.raw)},
+				RequestTime: formatTime(ping.request.t),
+			})
+			continue
+		}
+		var responses []*SinglePingResponse
+		for _, resp := range ping.responses {
+			responses = append(responses, makeResponse(resp))
+		}
+		tk.Pings = append(tk.Pings, &SinglePing{
+			ConnIdDst:   ping.request.dstID,
+			ConnIdSrc:   ping.request.srcID,
+			Failure:     nil,
+			Request:     &model.ArchivalMaybeBinaryData{Value: string(ping.request.raw)},
+			RequestTime: formatTime(ping.request.t),
+			Responses:   responses,
+		})
+	}
+	return nil
+}
+
+// DissectVersionNegotiation dissects the Version Negotiation response.
+// It returns the supported versions and the destination connection ID of the response,
+// The destination connection ID of the response has to coincide with the source connection ID of the request.
+// https://www.rfc-editor.org/rfc/rfc9000.html#name-version-negotiation-packet
+func (m *Measurer) DissectVersionNegotiation(i []byte) ([]uint32, ConnectionID, error) {
+	firstByte := uint8(i[0])
+	mask := 0b10000000
+	mask &= int(firstByte)
+	if mask == 0 {
+		return nil, nil, &errUnexpectedResponse{msg: "not a long header packet"}
+	}
+
+	versionBytes := i[1:5]
+	v := binary.BigEndian.Uint32(versionBytes)
+	if v != 0 {
+		return nil, nil, &errUnexpectedResponse{msg: "unexpected Version Negotiation format"}
+	}
+
+	dstLength := i[5]
+	offset := 6 + uint8(dstLength)
+	dst := i[6:offset]
+
+	srcLength := i[offset]
+	offset = offset + 1 + srcLength
+
+	n := uint8(len(i))
+	var supportedVersions []uint32
+	for offset < n {
+		supportedVersions = append(supportedVersions, binary.BigEndian.Uint32(i[offset:offset+4]))
+		offset += 4
+	}
+	return supportedVersions, dst, nil
+}
+
+// NewExperimentMeasurer creates a new ExperimentMeasurer.
+func NewExperimentMeasurer(config Config) model.ExperimentMeasurer {
+	return &Measurer{config: config}
+}
+
+// SummaryKeys contains summary keys for this experiment.
+//
+// Note that this structure is part of the ABI contract with probe-cli
+// therefore we should be careful when changing it.
+type SummaryKeys struct {
+	IsAnomaly bool `json:"-"`
+}
+
+// GetSummaryKeys implements model.ExperimentMeasurer.GetSummaryKeys.
+func (m *Measurer) GetSummaryKeys(measurement *model.Measurement) (interface{}, error) {
+	return SummaryKeys{IsAnomaly: false}, nil
+}
+
+// buildHeader creates the unprotected QUIC header.
+// https://www.rfc-editor.org/rfc/rfc9000.html#name-initial-packet
+func buildHeader(destConnID, srcConnID ConnectionID, payloadLen int) []byte {
```

Comment:
Please, move this function to a separate file. Logically, this is a basic building block for the main implementation of the experiment, so I'd like it to be in a separate file. The way I see it, our stack here is like:

```
  experiment (quicping.go)
  --------
  separate file (maybe quic.go?)
  --------
  crypto (crypto.go)

```